crypto/asn1/a_sign.c

```diff
@@ -144,6 +144,22 @@ int ASN1_item_sign_ex(const ASN1_ITEM *it, X509_ALGOR *algor1,
     return rv;
 }
 
+static int replace_algor_contents_from_DER(X509_ALGOR *algor, const unsigned char *aid, size_t aid_len)
+{
+    /* as a workaround for d2i_*() freeing its first argument, using NULL instead: */
+    X509_ALGOR *alg = d2i_X509_ALGOR(NULL, &aid, aid_len);
+
+    if (alg == NULL) {
+        ERR_raise(ERR_LIB_ASN1, ASN1_R_DECODE_ERROR);
+        return 0;
+    }
+    ASN1_OBJECT_free(algor->algorithm);
+    ASN1_TYPE_free(algor->parameter);
+    *algor = *alg;
+    OPENSSL_free(alg);
+    return 1;
+}
```

Comment:
I assume that you call OPENSSL_free(alg) and not X509_ALGOR_free(alg), to transfer the ownership of alg->algorithm and alg->parameter from alg to algor.